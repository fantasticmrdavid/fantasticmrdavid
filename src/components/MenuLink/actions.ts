import { Dispatch } from 'redux';
import { LOCATION_UPDATED } from 'reducers/location';
import {
  MEDIA_LOADING_STARTED,
  MEDIA_LOADING_STOPPED,
} from 'reducers/loading';

export const locationUpdated = (newLocation: string) => (dispatch: Dispatch<any>) => {
  dispatch({ type: MEDIA_LOADING_STARTED });
  dispatch({
    type: LOCATION_UPDATED,
    newLocation,
  });
  dispatch({ type: MEDIA_LOADING_STOPPED });
};
