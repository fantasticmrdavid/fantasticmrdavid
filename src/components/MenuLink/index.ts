import { withRouter } from 'react-router-dom';
import Ui from './ui';

export default withRouter(Ui);
