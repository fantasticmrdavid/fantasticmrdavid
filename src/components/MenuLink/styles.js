import styled from 'styled-components';
import theme from 'styles/theme';
import { media } from 'styles/utils';

const colorAbout = '#3fa44e';
const colorWork = '#967fd1';
const colorContact = '#cc750c';

export const Container = styled.div`
  display: block;
  position: relative;
  color: inherit;

  &.current {
    transition: 0s;
  }
`;

export const Label = styled.div`
  position: relative;
  padding: ${props => (props.menuActive ? '0.75rem 2rem' : '0.75rem 0')};
  transition: 0.5s;
  color: ${props => (props.current ? theme.colorBlack : undefined)};
  font-size: ${props => (props.current ? '2rem' : undefined)};
  font-weight: ${props => (props.current ? 600 : undefined)};
  cursor: pointer;
  background-color: ${(props) => {
    const { current, target } = props;
    if (current) {
      if (target === 'about') return colorAbout;
      if (target === 'work') return colorWork;
      if (target === 'contact') return colorContact;
      return theme.colorLightGrey;
    }
    return undefined;
  }};

  &:before {
    position: absolute;
    top: 0px;
    left: 0px;
    content: "";
    height: 100%;
    transition: 0.3s;
    background-color: ${theme.colorDarkGrey};
  }

  ${media.medium`
    &:hover {
      padding-left: ${props => (!props.current ? '60px' : undefined)};
      background-color: ${props => (!props.current ? theme.colorDarkGrey : undefined)};

      &:before {
        width: ${props => (props.current ? '0px' : '15px')};
        background-color: ${/* eslint-disable */
          (props) => {
          const { current, target } = props;
          if (!current) {
            if (target === 'about') return colorAbout;
            if (target === 'work') return colorWork;
            if (target === 'contact') return colorContact;
            return undefined;
          }
          return theme.colorDarkCharcoal;
        }/* eslint-enable */};
      }
    }
  `};
`;
