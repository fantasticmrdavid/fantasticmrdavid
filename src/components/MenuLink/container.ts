import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { AppState } from 'reducers';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { locationUpdated } from './actions';
import Ui, { Props as UiProps } from './ui';

type StateProps = {
  current: boolean,
}

type DispatchProps = {
  updateLocation: (target: string) => void,
}

type ContainerProps = UiProps & RouteComponentProps;

const mapStateToProps = (state: AppState, ownProps: UiProps) => {
  const { current, target } = ownProps;
  return {
    current: current || target === state.location.current,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<any>) => {
  return {
    updateLocation: (target: string) => {
      dispatch(locationUpdated(target));
    },
  };
};

export default connect<StateProps, DispatchProps, UiProps, AppState>(
  mapStateToProps,
  mapDispatchToProps,
)(withRouter(Ui));
