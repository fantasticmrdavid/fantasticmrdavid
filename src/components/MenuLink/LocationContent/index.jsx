import React, { memo } from 'react';
import PropTypes from 'prop-types';
import About from './About';
import Work from './Work';
import Contact from './Contact';
import * as styles from './styles';

const LocationContent = memo(
  ({ location, current }) => {
    const { Content } = styles;

    if (!location) return null;

    return (
      <Content current={current}>
        { location === 'about' && <About /> }
        { location === 'work' && <Work /> }
        { location === 'contact' && <Contact /> }
      </Content>
    );
  },
);

LocationContent.propTypes = {
  location: PropTypes.string.isRequired,
  current: PropTypes.bool,
};

export default LocationContent;
