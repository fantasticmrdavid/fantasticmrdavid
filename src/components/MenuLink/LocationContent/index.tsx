import React, { memo } from 'react';
import About from './About';
import Work from './Work';
import Contact from './Contact';
import * as styles from './styles';

interface Props {
  current: boolean,
  location: string,
}

export default memo(
  ({ location, current }: Props) => {
    const { Content } = styles;

    if (!location) return null;

    return (
      <Content current={current}>
        { location === 'about' && <About /> }
        { location === 'work' && <Work /> }
        { location === 'contact' && <Contact /> }
      </Content>
    );
  },
);
