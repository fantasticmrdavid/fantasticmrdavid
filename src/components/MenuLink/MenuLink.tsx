import React, { memo, SyntheticEvent, useContext } from 'react';
import { LocationContext } from 'contexts/Location';
import { LOCATIONS } from 'routers/constants';
import {
  Container,
  Label,
} from './menuLink.styles';
import { LocationContent } from './LocationContent/LocationContent';
import { useNavigate } from 'react-router-dom';

export interface MenuLinkProps {
  target: typeof LOCATIONS[keyof typeof LOCATIONS],
  isCurrent: boolean,
  isMenuActive: boolean,
  label: string,
}

export const MenuLink = memo(
  (props: MenuLinkProps) => {
    const {
      isCurrent,
      label,
      target,
      isMenuActive,
    } = props;

    const navigate = useNavigate();

    const { setLocation } = useContext(LocationContext);

    const handleNavClick = (e: SyntheticEvent) => {
      e.preventDefault();
      e.stopPropagation();
      setLocation(isCurrent ? LOCATIONS.home : target);
      navigate(`/${isCurrent ? '' : target}`);
    };

    return (
      <Container>
        <Label
          isCurrent={isCurrent}
          isMenuActive={isMenuActive}
          target={target}
          onClick={handleNavClick}
        >
          { label }
        </Label>
        <LocationContent location={target} isCurrent={isCurrent} />
      </Container>
    );
  },
);
