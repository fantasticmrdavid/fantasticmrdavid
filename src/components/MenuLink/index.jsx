import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import LocationContent from './LocationContent';
import styles from './styles.css';

class MenuLink extends PureComponent {
  constructor() {
    super();
    this.boundHandleNavClick = this.handleNavClick.bind(this);
  }

  handleNavClick(e) {
    const { target, current, updateLocation } = this.props;
    e.preventDefault();
    e.stopPropagation();
    updateLocation(current ? 'home' : target);
  }

  render() {
    const { label, target, current, menuActive } = this.props;
    return (
      <div className={`${styles.MenuLink} ${current && styles.current} ${menuActive && styles.menuActive}`} onClick={this.boundHandleNavClick}>
        <div className={`${styles.label} ${styles[target]}`} data-target={target} onClick={this.boundHandleNavClick}>{ label }</div>
        <LocationContent location={target} current={current} />
      </div>
    );
  }
}

MenuLink.propTypes = {
  updateLocation: PropTypes.func.isRequired,
  target: PropTypes.string.isRequired,
  current: PropTypes.bool,
  menuActive: PropTypes.bool,
  label: PropTypes.string,
};

export default MenuLink;
