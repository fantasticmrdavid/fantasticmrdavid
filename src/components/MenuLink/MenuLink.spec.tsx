import React from 'react';
import { render } from "@testing-library/react";
import { MenuLink } from 'components/MenuLink/MenuLink';
import { LOCATIONS } from 'routers/constants';

const props = {
  target: LOCATIONS.contact,
  isCurrent: false,
  isMenuActive: true,
  label: 'Contact',
  history: {
    push: jest.fn(),
  },
};

describe('Component -> MenuLink', () => {
  it('should render the "contact" MenuLink UI without throwing an error', () => {
    expect(render(<MenuLink {...props} />)).toMatchSnapshot();
  });

  it('should render the "work" MenuLink UI without throwing an error', () => {
    expect(render(<MenuLink {...props} target="work" />)).toMatchSnapshot();
  });

  it('should render the "about" MenuLink UI without throwing an error', () => {
    expect(render(<MenuLink {...props} target="about" />)).toMatchSnapshot();
  });
});
