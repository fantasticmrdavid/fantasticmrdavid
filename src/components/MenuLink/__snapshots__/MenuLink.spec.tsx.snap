// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component -> MenuLink should render the "about" MenuLink UI without throwing an error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="menuLinkstyles__Container-sc-1da8bov-0 gwazlq"
      >
        <div
          class="menuLinkstyles__Label-sc-1da8bov-1 lneLWJ"
          target="about"
        >
          Contact
        </div>
        <div
          class="locationContentstyles__Content-sc-osxrd6-0 YmcJA"
        >
          <div>
            Meet the captain of the ship.
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="menuLinkstyles__Container-sc-1da8bov-0 gwazlq"
    >
      <div
        class="menuLinkstyles__Label-sc-1da8bov-1 lneLWJ"
        target="about"
      >
        Contact
      </div>
      <div
        class="locationContentstyles__Content-sc-osxrd6-0 YmcJA"
      >
        <div>
          Meet the captain of the ship.
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Component -> MenuLink should render the "contact" MenuLink UI without throwing an error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="menuLinkstyles__Container-sc-1da8bov-0 gwazlq"
      >
        <div
          class="menuLinkstyles__Label-sc-1da8bov-1 fsWMbj"
          target="contact"
        >
          Contact
        </div>
        <div
          class="locationContentstyles__Content-sc-osxrd6-0 YmcJA"
        >
          <div>
            For relaxing times, make it Suntory time.
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="menuLinkstyles__Container-sc-1da8bov-0 gwazlq"
    >
      <div
        class="menuLinkstyles__Label-sc-1da8bov-1 fsWMbj"
        target="contact"
      >
        Contact
      </div>
      <div
        class="locationContentstyles__Content-sc-osxrd6-0 YmcJA"
      >
        <div>
          For relaxing times, make it Suntory time.
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Component -> MenuLink should render the "work" MenuLink UI without throwing an error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="menuLinkstyles__Container-sc-1da8bov-0 gwazlq"
      >
        <div
          class="menuLinkstyles__Label-sc-1da8bov-1 dPyKiY"
          target="work"
        >
          Contact
        </div>
        <div
          class="locationContentstyles__Content-sc-osxrd6-0 YmcJA"
        >
          <div>
            Some things I've been working on.
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="menuLinkstyles__Container-sc-1da8bov-0 gwazlq"
    >
      <div
        class="menuLinkstyles__Label-sc-1da8bov-1 dPyKiY"
        target="work"
      >
        Contact
      </div>
      <div
        class="locationContentstyles__Content-sc-osxrd6-0 YmcJA"
      >
        <div>
          Some things I've been working on.
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
